Persist Prerendered State

surayya-MS opened the pull request:
# Persist Prerendered State

## Description

https://learn.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/persist-component-state?view=aspnetcore-7.0

With this PR the usage example stays exactly the same as in the link above.





Fixes https://github.com/dotnet/aspnetcore/issues/49733

surayya-MS committed:
do not clear store when circuit is initialized

surayya-MS committed:
provide a way to persist data for Blazor web

surayya-MS committed:
small fix

surayya-MS committed:
fix tests

surayya-MS committed:
update application state when document is updated

surayya-MS committed:
provide summaries

surayya-MS committed:
merge from main

surayya-MS committed:
small fixes

surayya-MS committed:
fix updating persisted state when document is updated

surayya-MS commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -46,9 +55,43 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
+            throw new InvalidOperationException("Cannot infer serialization mode for non component. Provide a serialization mode.");
+        }
+
+        var serializationMode = SerializationMode;
```
When the code is executed on the circuit (or webassembly runtime) there is no way to get the render mode of the component. For this reason I changed [ComponentStatePersistenceManager.RestoreStateAsync](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-0386c940190e808072229a79b79b845e646152920641a931b44e619ceb4f1efcR51) to provide the serialization mode.

surayya-MS replied:
Thanks for the suggestion! I tried it out and it won't work. 
The reason is that `ComponentStatePersistenceManager.PersistStateAsync(store, serializationMode, dispatcher)` persists the state for the provided serializationMode (`Server` or `WebAssembly`) into the provided store. We want to store server state and webassembly state into differrent stores. Currently we store the state in `ProtectedPrerenderComponentApplicationStore` for Blazor Server and `PrerenderComponentApplicationStore` for Blazor WebAssembly.

That is why it seemed like a good idea at the time to change `ComponentStatePersistenceManager.RestoreStateAsync(store, serializationMode)` by adding serializationMode argument. We persist the state for the given serialization mode in the given store and similarly restoring the the state for serialization mode from the given store.
But I agree that in the implementation `RestoreStateAsync` doesn't do what you'd expect and indeed ,as you said, it relies on the fact that RestoreStateAsync always gets called before RegisterOnPersisting.

I found a way to fix it. I added `GlobalSerializationMode` in [ISerializationModeHandler](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-b58ef65f1a58ba6b1c30959855cf80b9d8d935681fdc98ed28cb719086b338e6R21). I change the `GlobalSerializationMode` before I call `RestoreStateAsync`. 
`GlobalSerializationMode` is also useful in the case when the code is running on the circuit(or webassembly runtime) and you try to register a callback on webassembly(server). The callback should not be fired. By checking the `GlobalSerializationMode` I throw the exception.

MackinnonBuck replied:
> Thanks for the suggestion! I tried it out and it won't work.
> The reason is that ComponentStatePersistenceManager.PersistStateAsync(store, serializationMode, dispatcher) persists the state for the provided serializationMode (Server or WebAssembly) into the provided store. We want to store server state and webassembly state into differrent stores. 

Yeah, that makes sense - it looks like I didn't cover that detail in my initial comment here. I sketched out some of the ideas from my first comment [as a commit](https://github.com/dotnet/aspnetcore/commit/f7a8552d1de5076e123f48f7cf06f8a4224ac33e), because code is probably a lot easier to parse than a list of bullet points 😄

What do you think? The goal with the ideas I'm proposing is to reduce the amount of code dedicated to tracking Server/WebAssembly-specific state. Feel free to borrow some of those ideas if you agree with them (or ignore them if you prefer what you have).

Your approach using a `GlobalSerializationMode` property is also valid - my only concern with that is it's more difficult to reason about how configuration of global state impacts the behavior of the program. Ideally, I think it would be nice to have most of the logic be agnostic to the specific serialization mode.

surayya-MS commented on src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs:
```diff
@@ -34,105 +46,94 @@ public ComponentStatePersistenceManager(ILogger<ComponentStatePersistenceManager
                 try
                 {
-                    await task;
+                    await callback();
```
I made the execution of the callbacks sequential here. The reason is that `PersistAsJson` is usually provided in the callback. In order to persist the key, value in the correct dictionary `PersistAsJson` uses the `SerializaionState` property which can be set only here before foreach block. If I'd leave the execution of the callbacks parallel the `SerializaionState` can be set incorrectly.

MackinnonBuck replied:
I think running callbacks in parallel should still be possible, as long as we wait for the completion of each callback before mutating `SerializationMode` again.

> If I'd leave the execution of the callbacks parallel the SerializaionState can be set incorrectly.

Do you remember why exactly this was happening?

surayya-MS replied:
You are right. Changed the code to await the completion of all callbacks.

surayya-MS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -12,12 +12,21 @@ export function discoverComponents(root: Node, type: 'webassembly' | 'server' |
 }
 
-const blazorStateCommentRegularExpression = /^\s*Blazor-Component-State:(?<state>[a-zA-Z0-9+/=]+)$/;
+const blazorServerStateCommentRegularExpression = /^\s*Blazor-Server-Component-State:(?<state>[a-zA-Z0-9+/=]+)$/;
```
We need to discover server and webassembly states.

MackinnonBuck commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -46,9 +55,43 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
         ArgumentNullException.ThrowIfNull(callback);
 
-        _registeredCallbacks.Add(callback);
+        if (callback.Target is not IComponent)
```
Really cool idea!!

surayya-MS replied:
@javiercn's idea

MackinnonBuck commented on src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs:
```diff
@@ -33,19 +35,45 @@ protected override IComponent ResolveComponentForRenderMode([DynamicallyAccessed
 
+    public PersistedStateSerializationMode GetComponentSerializationMode(IComponent component)
+    {
+        var ssrRenderBoundary = GetClosestRenderModeBoundary(component);
+        var renderMode = ssrRenderBoundary!.RenderMode;
```
What happens if there's no render mode boundary? This can happen if a render mode wasn't set in an ancestor component.

surayya-MS replied:
Thanks for noting! I wanted to get back to it and throw an error and forgot. Fixed now

MackinnonBuck commented on src/Components/Web.JS/src/Services/WebRootComponentManager.ts:
```diff
@@ -164,6 +165,21 @@ export class WebRootComponentManager implements DescriptorHandler, NavigationEnh
     }
   }
 
+    private updateApplicationState() {
+      if (isRendererAttached(WebRendererId.Server)) {
+        const serverAppState = discoverServerPersistedState(document);
+        if (serverAppState) {
+            updateApplicationState(WebRendererId.Server, serverAppState);
+        }
+      }   
+      if (isRendererAttached(WebRendererId.WebAssembly)) {
+        const webAssemblyAppState = discoverWebAssemblyPersistedState(document);
+        if (webAssemblyAppState) {
+            updateApplicationState(WebRendererId.WebAssembly, webAssemblyAppState);
+        }
+      }  
+    }
```
I wonder if there are things that we can do to avoid scanning the entire DOM each time the document receives an SSR update.

With component comments, we scan the incoming DOM before it gets merged with the document. That allows us to only scan the subset of the document that gets changed. Could we consider a similar approach here?

surayya-MS replied:
Agree. Working on it

MackinnonBuck reviewed (commented):
Looks like a great start!

surayya-MS committed:
fire callbacks in parallel

surayya-MS committed:
chek for a component in the interface method

surayya-MS committed:
fix tests

surayya-MS committed:
changed ISerializationModeHandler

surayya-MS committed:
implemented only server and only webassembly ISerializationModeHandler

surayya-MS commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -46,9 +55,43 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
-        _registeredCallbacks.Add(callback);
+        if (callback.Target is not IComponent)
+        {
+            throw new InvalidOperationException("Cannot infer serialization mode for non component. Provide a serialization mode.");
```
Good point, thanks for noticing! 

I changed the code by moving the check for callback target being a component into the `EndpointHtmlRenderer.GetCallbackTargetSerializationMode`. 

I also implemented `ServerSerializationModeHandler` and `WebAssemblySerializationModeHandler` for pure Blazor Server and Blazor WebAssembly. 

I wasn't sure where to add the `ServerSerializationModeHandler`. Seems like this is a right place https://github.com/dotnet/aspnetcore/pull/50373/files#diff-c1d00e1ea2b0628c9c2d2a32c1e59e625dd3c972e9938b4b0f33d2709e9ec5bfR212

surayya-MS commented on src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs:
```diff
@@ -34,105 +46,94 @@ public ComponentStatePersistenceManager(ILogger<ComponentStatePersistenceManager
-    public Task PersistStateAsync(IPersistentComponentStateStore store, Dispatcher dispatcher)
+    public Task PersistStateAsync(
+        IPersistentComponentStateStore store,
+        PersistedStateSerializationMode serializationMode,
```
Explained in this [comment](https://github.com/dotnet/aspnetcore/pull/50373#discussion_r1313171647)

MackinnonBuck replied:
I sketched out some of my ideas in  https://github.com/dotnet/aspnetcore/commit/f7a8552d1de5076e123f48f7cf06f8a4224ac33e to show how this could be achieved 🙂

surayya-MS committed:
fire the callbacks in parallel

surayya-MS committed:
merge from main

surayya-MS committed:
small fixes

surayya-MS committed:
fix tests

MackinnonBuck committed:
Some ideas

javiercn commented on src/Components/Components/src/ISerializationModeHandler.cs:
```diff
@@ -0,0 +1,23 @@
+/// <summary>
+/// A service that can infer <see cref="IComponent"/>'s <see cref="PersistedStateSerializationMode"/>.
+/// </summary>
+public interface ISerializationModeHandler
```
Why do we need this interface? Can this be done in a different way?

surayya-MS replied:
The reason for creating this interface is that it was impossible to reference `EndpointHtmlRenderer` in `PersistentComponentState`. Alternative solution was to add some render mode specific absctract methods to the `Renderer` and then override them in `EndpointHtmlRenderer` which didn't feel right.

javiercn replied:
>The reason for creating this interface is that it was impossible to reference EndpointHtmlRenderer in PersistentComponentState.

Have you checked-out Steve's spike? It doesn't require those.

> Alternative solution was to add some render mode specific absctract methods to the Renderer and then override them in EndpointHtmlRenderer which didn't feel right.

That sounds desirable over a separate interface that needs to be implemented in many places.

surayya-MS replied:
`ISerializationModeHandler` also solves the issue when the code start to run on the circuit or webassembly runtime. When this happens render mode for the components is not available. The reason is that circuit/webassembly runtime uses it own scope and the Renderer's dictionary for component states is not populated. More details this in [this comment](https://github.com/dotnet/aspnetcore/pull/50373#discussion_r1318463261)

javiercn commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -13,26 +13,36 @@ namespace Microsoft.AspNetCore.Components;
 public class PersistentComponentState
 {
     private IDictionary<string, byte[]>? _existingState;
```
Why do we need _existingState anymore?

surayya-MS replied:
We need `_existingState` for restoring the application state after emitting the comment containing the application state into html page.

javiercn commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -13,26 +13,36 @@ namespace Microsoft.AspNetCore.Components;
 
     internal bool PersistingState { get; set; }
 
+    internal PersistedStateSerializationMode CurrentSerializationMode { get; set; } = PersistedStateSerializationMode.Infer;
```
Why is the default "infer"?

surayya-MS replied:
It doesn't need to be. I'll remove this. Thanks for noting!

javiercn commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -46,9 +56,38 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
+        ArgumentNullException.ThrowIfNull(callback);
+
+        if (_serializationModeHandler.GlobalSerializationMode != PersistedStateSerializationMode.Infer &&
+            serializationMode != _serializationModeHandler.GlobalSerializationMode)
```
Can you enumerate the possible combinations for the different use cases? It's not clear how this works E2E

surayya-MS replied:
This means "Unless the GlobalSerializationMode is not infer you cannot persist state for serializationMode different that the GlobalSerializationMode".
I'll provide a clarifying comment, since it raised a question.

javiercn commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -60,19 +99,26 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
+        {
+            PersistedStateSerializationMode.Server => _currentServerState,
+            PersistedStateSerializationMode.WebAssembly => _currentWebAssemblyState,
+            _ => throw new InvalidOperationException("Invalid persistence mode.")
```
How does this work with AutoMode? One of the key points of the design was that calls to `PersistAsJson` still work without specifying the "serialization mode" when the callback comes from a component.

How does that work in this scenario? It's throwing isn't it?

surayya-MS replied:
Haven't tried this scenario yet.

javiercn commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -60,19 +99,26 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
+            _ => throw new InvalidOperationException("Invalid persistence mode.")
+        };
+
         ArgumentNullException.ThrowIfNull(key);
```
This should happen before any other check

javiercn commented on src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblySerializationModeHandler.cs:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    public PersistedStateSerializationMode GetCallbackTargetSerializationMode(object? callbackTarget)
+        => PersistedStateSerializationMode.WebAssembly;
```
I'm failing to understand what this is doing if it's just returning a constant instead of looking at the callback target.

surayya-MS replied:
Well, short answer to your question is because for webassembly the callback target can only be `PersistedStateSerializationMode.WebAssembly`.

I guess the confusing part is why there are 3 `ISerializationModeHandler`s: `EndpointHtmlRenderer`, `ServerSerializationModeHandler`, `WebAssemblySerializationModeHandler` (actually now listing them, I realised it might be better to add `EndpointSerializationModeHandler` and use `EndpointHtmlRenderer` inside of it).

surayya-MS replied:
Well, short answer to your question is because for webassembly the callback target can only be `PersistedStateSerializationMode.WebAssembly`.

I guess the confusing part is why there are 3 `ISerializationModeHandlers`: `EndpointHtmlRenderer`, `ServerSerializationModeHandler`, `WebAssemblySerializationModeHandler` (actually now listing them, I realised it might be better to add `EndpointSerializationModeHandler` and use `EndpointHtmlRenderer` inside of it).

**EndpointHtmlRenderer**
`GetCallbackTargetSerializationMode` checks if the callback target is a component and then gets the component's render mode. 
However this doesn't work on the circuit because the component's render mode is not available on the circuit. The reason is that the circuit doesn't use the same instance of `PersistentComponentState` and `ISerializationModeHandlers`, it has its own scope. The circuit uses same exact kinds of services listed in `AddRazorComponents` though.
Hence, there is a need to be able to change the `GlobalSerializationMode` to `Server` when creating the circuit host [here](https://github.com/dotnet/aspnetcore/pull/50373/files/7bbfde083b1f483cd2fc3686cbec1e1f41420281#diff-4a5e6027a7a100cacb4a0f69d0c548e7c01379898af20a265297840d171f38b0R66).


**WebAssemblySerializationModeHandler**
The webassembly host uses the services provided [here](https://github.com/dotnet/aspnetcore/pull/50373/files/7bbfde083b1f483cd2fc3686cbec1e1f41420281#diff-e5845021124dd8ca74320ce8dbc62fa89ed95f3a7addf44ed99120a9d501a279R254). 
Thant is why I created `WebAssemblySerializationModeHandler` that will just return constant `PersistedStateSerializationMode.WebAssembly` and throws when trying to set `GlobalSerializationMode`.


**ServerSerializationModeHandler**
This one is for pure Blazor Server apps. I baked it [here](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-c1d00e1ea2b0628c9c2d2a32c1e59e625dd3c972e9938b4b0f33d2709e9ec5bfR212) before `AddRazorComponents` so it would override `EndpointHtmlRenderer`. This seemed like a right place after following AddRazor**Pages**.


Hope this answers your question.

javiercn commented on src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.PrerenderingState.cs:
```diff
@@ -15,6 +15,16 @@ internal partial class EndpointHtmlRenderer
 {
     private static readonly object InvokedRenderModesKey = new object();
 
+    public async ValueTask<IHtmlContent> PrerenderServerPersistedStateAsync()
```
Why do we need to go through IHtmlContent for this? Seems like unnecessary overhead.

surayya-MS replied:
I don't think I understand your question, why do you think this might be unnecessary overhead? This is similar to the existing [ValueTask IHtmlContent PrerenderPersistedStateAsync(HttpContext httpContext, PersistedStateSerializationMode serializationMode)](https://github.com/dotnet/aspnetcore/pull/50373/files/7bbfde083b1f483cd2fc3686cbec1e1f41420281#diff-ed60d57e506a39f7e43eacf616f877964403dba2bc6626a7638e2cbee1a41057R28-R29)

javiercn replied:
Couldn't we just write directly to the response?

javiercn reviewed (commented):
This is a step in the good direction. However, I am not sure this design correctly aligns with the discussions that we previously had.

It's unclear to me how this design works when there are server and Webassembly components running on the same app. 

There is a lack of E2E tests on the PR that make it hard to understand/validate what scenarios work and how.

It's unclear to me why we need an `ISerializationModeHandler` abstraction, when this was something that we were able to handle completely internally in the past.

I am not grocking the `GlobalSerializationMode`. In my head there is no such concept, as the serialization mode is local to the component render mode or service.

Server render mode component -> Server
Webassembly render mode component -> Webassembly.
Auto render mode component -> Server and webassembly.
DI service when only webassembly is enabled -> Webassembly
DI service when only server is enabled -> Server
DI service when both server and webassembly is enabled -> Server.

Explicit serialization options:
* ServerOnly
* WebAssemblyOnly
* All

Overall, I think that the serialization/deserialization should be kept internal to the `EndpointHtmlRenderer`.

The only thing that ComponentState needs to know is the render mode (if any) for a given callback, which is something that the renderer can provide.

If it doesn't know the render mode, it can throw or it can fallback to something safe (like Server), and we can provide an option to change the default (we can even do this on a per-type basis).

SteveSandersonMS commented:
Great to see this, @surayya-MS!

I know you might have an excess of review feedback by now since @MackinnonBuck and @javiercn have already given comments, so I'll try to keep this super focused!

My main feedback is that I would recommend taking an approach like @MackinnonBuck described [here](https://github.com/dotnet/aspnetcore/pull/50373#discussion_r1309240490), though maybe with a few extra tweaks to minimize new public API. That, is:

 * Have `RegisterOnPersisting` take an extra optional `IComponentRenderMode` and store this inside `PersistentComponentState`'s `_registeredCallbacks`
 * Also have `ComponentStatePersistenceManager.PersistStateAsync` take an optional extra "filter" parameter, e.g., of type `public delegate bool PersistentStateCallbackFilter(object? target, IComponentRenderMode? renderMode);`
   * It can get the `IComponentRenderMode` from `_registeredCallbacks`, and `target` from `Callback.Target`
 * Then `EndpointHtmlRenderer`'s `PrerenderPersistedStateAsync` can have two overloads:
   * One that still takes a `PersistedStateSerializationMode` and has the same logic as before. The "filter" parameter can just return true for everything, since `PersistedStateSerializationMode` controls everything. This continues to be used for back-compat by the `<persist-component-state>` tag helper.
  * A new one that doesn't take `PrerenderPersistedStateAsync`, but instead serialized both server and webassembly state by using two different stores, and "filter" callbacks that do something like:
    * If there's a `renderMode` on the registration, match based on that
    * If not, but if `target` is a component, look up the effective rendermode for `target` and match based on that
    * Else just omit the value completely (or, equally good, if we're serializing server state, include nonmatched items by default)

This is almost the same as Mackinnon's suggestion except it leaves the renderer in control of how to interpret `IComponentRenderMode` and avoids baking in the concept of "server vs WebAssembly" at the M.A.Components level which is meant to be agnostic to those concepts. This also eliminates the need for a "global serialization mode" concept.

I did mock up this approach and it appeared to work (let me know if it's useful to see that), but I know you've dug deeper into this area already so apologies if I'm missing some details.

surayya-MS committed:
fix PublicAPI

surayya-MS committed:
1. Add ServerAndWebAssembly PersistedStateSerializationMode
2. Fixed PrerenderPersistedStateAsync
3. Moved the check _stateIsPersisted from ComponentStatePersistenceManager to the store
4. Made possible to change ISerializationModeHandler to Server one for circuit

surayya-MS committed:
fix

surayya-MS commented:
@MackinnonBuck, thanks for taking time to code your ideas! It is much easier. I really like that the `PersistentComponentState` and `ComponentStatePersistenceManager` look much cleaner! Also, cool idea to add `SupportsSerializationMode` to `IPersistentComponentStateStore`! I included your changes. Thanks a lot!

However, I had to change a few things. 
1. When code runs on the circuit there is no serialization mode available. Same problem as before with adding extra serializationMode to `RestoreAsync` or changing `GlobalSerializationMode`. This time I added [ComponentStatePersistenceManager.SetSerializationModeHandler](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-0386c940190e808072229a79b79b845e646152920641a931b44e619ceb4f1efcR36). I change the `ISerializationMode` to `ServerSerializationMode` when creating the circuit host [here](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-4a5e6027a7a100cacb4a0f69d0c548e7c01379898af20a265297840d171f38b0R65).
Also I'm concerned about the case when the code runs on circuit(or webassembly) and someone registers a callback with webassembly(or server) serialization mode. With the current design we don't throw exeption. The callback is just never fired. Are we okay with this @SteveSandersonMS , @javiercn ? With the `GlobalSerializationMode` design it was possible. Bringing it back also doesn't make sense since it will bring back the complexity of inferring the serialization mode back to `PersistenComponentState`. I'm fine with the current code and to throw an exception. We should just mention this in the docs.

2. [Inferring the serialization mode from InvokedRenderModes](https://github.com/dotnet/aspnetcore/commit/f7a8552d1de5076e123f48f7cf06f8a4224ac33e#diff-ed60d57e506a39f7e43eacf616f877964403dba2bc6626a7638e2cbee1a41057R26-R31) doesn't work. It always `InvokedRenderModes.Mode.None`. For that reason I added [PersistedStateSerializationMode.ServerAndWebAssembly](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-17c51ec93416d656ccefe736f2b468e50b1a075bb3b6fcda0a9a8eeb0a75a7f5R29).
This change makes the code shorter in case when someone wants to register the callback for both server and webassembly. Also had to throw exception when serialization mode is infer. 

3. The exception is thrown saying that the state is already persisted. This happens when `ComponentStatePersistenceManager.PersistStateAsync` is called for the second time. For this reason I moved this check inside the [PrerenderComponentApplicationStore.PersistStateAsync](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-b89efe35765a7c673da1e1c26a4072f77d2f6467e7a4db8de143768efe854af3R59-R64)

javiercn commented on src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs:
```diff
@@ -44,47 +52,48 @@ public async Task RestoreStateAsync(IPersistentComponentStateStore store)
+            var currentState = new Dictionary<string, byte[]>();
 
-            await store.PersistStateAsync(_currentState);
+            State.PersistenceContext = new(currentState);
+            await PauseAsync(store);
+            State.PersistenceContext = default;
```
We can avoid all this if we do things a bit differently.

* Collect or infer the mode inside the call to register for each callback.
* Persist components with Server mode in a batch.
* Persist components with Webassembly mode afterwards.

javiercn commented on src/Components/Components/src/PersistenceCallback.cs:
```diff
@@ -0,0 +1,6 @@
+
+namespace Microsoft.AspNetCore.Components;
+
+internal readonly record struct PersistenceCallback(Func<Task> Callback, PersistedStateSerializationMode SerializationMode);
```
This is only used inside PersistentComponentState, it's better to nest it inside that type. Having it as a top-level type adds unnecessary noise to the Microsoft.AspNetCore.Components namespace.

javiercn commented on src/Components/Components/src/PersistenceContext.cs:
```diff
@@ -0,0 +1,6 @@
+
+namespace Microsoft.AspNetCore.Components;
+
+internal readonly record struct PersistenceContext(IDictionary<string, byte[]> State);
```
Same comment as above, these types that are used just for structured data should be inside the class that uses them. You can also dispense creating a separate type when all you are holding is a single piece of data.

javiercn commented on src/Components/Components/src/PersistentComponentState.cs:
```diff
@@ -46,9 +46,26 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
+    /// <param name="callback">The callback to invoke when the application is being paused.</param>
+    /// <param name="serializationMode">The <see cref="PersistedStateSerializationMode"/> to register the callback.</param>
+    /// <returns>A subscription that can be used to unregister the callback when disposed.</returns>
+    public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> callback, PersistedStateSerializationMode serializationMode)
```
This should probably be `IComponentRenderMode` instead of `PersistedStateSerializationMode` since that's what we are telling users to use.

javiercn commented on src/Components/Endpoints/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs:
```diff
@@ -54,6 +54,7 @@ public static IRazorComponentsBuilder AddRazorComponents(this IServiceCollection
         services.TryAddSingleton<WebAssemblyComponentSerializer>();
         services.TryAddScoped<EndpointHtmlRenderer>();
         services.TryAddScoped<IComponentPrerenderer>(services => services.GetRequiredService<EndpointHtmlRenderer>());
+        services.TryAddScoped<ISerializationModeHandler>(services => services.GetRequiredService<EndpointHtmlRenderer>());
```
This is where the issue is. There is a hidden circular dependency here.

Renderer -> PersistentComponentStateManager -> Renderer(ISerializationModeHandler).

The thing that initiates the persistence process (the renderer) calls the `ComponentPersistenceManager` and has access to the renderer, which in turn can use that to determine the correct render mode https://github.com/dotnet/aspnetcore/blob/8eaf4b51f73ae2b0ed079e4f8253afb53e96b703/src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs#L50

javiercn commented on src/Components/Web/src/PublicAPI.Unshipped.txt:
```diff
@@ -121,6 +121,7 @@ static Microsoft.AspNetCore.Components.Web.RenderMode.WebAssembly.get -> Microso
 virtual Microsoft.AspNetCore.Components.HtmlRendering.Infrastructure.StaticHtmlRenderer.RenderChildComponent(System.IO.TextWriter! output, ref Microsoft.AspNetCore.Components.RenderTree.RenderTreeFrame componentFrame) -> void
 virtual Microsoft.AspNetCore.Components.HtmlRendering.Infrastructure.StaticHtmlRenderer.WriteComponentHtml(int componentId, System.IO.TextWriter! output) -> void
 virtual Microsoft.AspNetCore.Components.RenderTree.WebRenderer.GetWebRendererId() -> int
+virtual Microsoft.AspNetCore.Components.RenderTree.WebRenderer.UpdateApplicationState(string! applicationState) -> System.Threading.Tasks.Task!
```
Why is this needed at all? Previous versions didn't require it, and I don't see why this would be something that is required now.

surayya-MS replied:
With the new blazor web we prerender every page. This allows to update the application state. We discussed it with Mackinnon and Steve on the email thread

javiercn replied:
Can't find the email thread, can you point me to it?

surayya-MS replied:
forwarded the email to you

javiercn reviewed (commented):
Thanks for pushing through this @surayya-MS. I think we still need to iron out some of the details in the design/implementation. Specifically:
* We shouldn't need to add an ISerializationModeHandler public API bit to deal with inferring the persistence. [ComponentStatePersistenceManager](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs) already has access to the renderer and can use that  (though an internal method) to get the render mode for the component.
* UpdateApplicationState shouldn't need to be public. We've always restored the state for the renderer in the past in a private manner, and I don't see a reason why that needs to change.

Lastly, it's really hard to further evaluate the PR without the E2E scenarios that showcase the different scenarios and how they work, so I would strongly encourage you to add E2E scenarios.

surayya-MS commented:
>UpdateApplicationState shouldn't need to be public. We've always restored the state for the renderer in the past in a private manner, and I don't see a reason why that needs to change.

@javier, we restored the state in the past in a private manner when creating circuit host/wasm host. When the host already exists there needs to be a way to update the application state from the js side. UpdateApplication state solves that

javiercn commented:
> @javiercn, we restored the state in the past in a private manner when creating circuit host/wasm host. When the host already exists there needs to be a way to update the application state from the js side. UpdateApplication state solves that

That doesn't mean there needs to be a public API to do it.
